src/provider.tsx

```diff
@@ -0,0 +1,160 @@
+  auth0Client: Auth0Client,
+) => async (url?: string): Promise<RedirectLoginResult> => {
+  dispatch(handleRedirectCallbackAction.started());
+
+  const response = await auth0Client.handleRedirectCallback(url);
+  const user = await auth0Client.getUser();
+  log("handleRedirectCallback(): got response:", response);
+  log("handleRedirectCallback(): got user:", user);
+
+  dispatch(handleRedirectCallbackAction.done({ result: { user } }));
+
+  return response;
+};
+
+const buildLoginWithPopup = (dispatch: Dispatch<Auth0ProviderAction>, auth0Client: Auth0Client) => async (
+  options?: PopupLoginOptions,
+  config?: PopupConfigOptions,
+): Promise<void> => {
+  dispatch(loginWithPopupAction.started());
+
+  try {
+    await auth0Client.loginWithPopup(options, config);
+  } catch (error) {
+    console.error(error);
+  }
+
+  const user = await auth0Client.getUser();
+
+  dispatch(loginWithPopupAction.done({ result: { user } }));
+};
+
+export const Auth0Provider: React.FC<Auth0ProviderProps> = ({
+  children,
+  history,
+  enableDebugLogging,
+  renderLoading,
+  ...initOptions
+}) => {
+  const log = useLog(enableDebugLogging);
+
+  const [state, dispatch] = useReducer<Reducer<Auth0ProviderState, Auth0ProviderAction>>(
+    auth0ProviderStateReducer,
+    InitialReducerState,
+  );
+
+  const location = useLocation();
+
+  const { loading: initLoading, value: initValue } = useAsync(async () => {
+    log("Beginning auth0 initialization");
+
+    dispatch(auth0Loaded.started());
+
+    const auth0Client = await createAuth0Client(initOptions);
+
+    const handleRedirectCallback = buildHandleRedirectCallback(dispatch, log, auth0Client);
+
+    const loginWithPopup = buildLoginWithPopup(dispatch, auth0Client);
+
+    // if (location.search?.includes("code=") && location.search?.includes("state=")) {
+    //   log("Handling redirect");
+    //   const { appState } = await handleRedirectCallback();
+    //   onRedirectCallback(appState);
+    // }
+
+    const isAuthenticated = await auth0Client.isAuthenticated();
+
+    const user: TokenUser | undefined = (isAuthenticated && (await auth0Client.getUser())) || undefined;
+
+    return {
+      user,
+      auth0Client,
+      handleRedirectCallback,
+      loginWithPopup,
+    };
+  }, []);
+
+  const [handleRedirAsyncState, handleRedirAsync] = useAsyncFn(async () => {
+    if (!isLoadedState(state)) {
+      return;
+    }
+
+    log("Handling redirect");
+    const { appState: s } = await state.handleRedirectCallback();
+    const appState: RouteState = s;
+    log("Back from redirect, replacing state to ", appState);
+    history.replace(appState.targetUrl);
+  }, ["handleRedirectCallback" in state && state.handleRedirectCallback, log, history.replace]);
+
+  useEffect(() => {
+    if (!isLoadedState(state) || handleRedirAsyncState.loading) {
+      return;
+    }
+
+    if (location.search?.includes("code=") && location.search?.includes("state=")) {
+      handleRedirAsync().then(() => {
```

Comment:
do you need the `.then` at all, if it is just a noop

Reply:
Not really -- honestly, this is more to satisfy my IDE from bugging me that I'm not handling the result of a Promise.  I'm fine taking it out, although I think it does help indicate to the reader that something async happens here.  (Although the method name probably indicates it too ;))